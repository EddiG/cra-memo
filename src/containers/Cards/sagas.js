import { delay } from 'redux-saga';
import { take, race, select, call, put, takeEvery } from 'redux-saga/effects';
import { TRIGGER_CARD_ASYNC, VALIDATE_CARDS } from './constants';
import { triggerCard } from './actions';
import { selectOpenedNotGuessedCards } from './selectors';

export function* validateCardsAsync() {
  yield call(delay, 1000);
  yield put({ type: VALIDATE_CARDS });
}

export function* triggerCardAsync(action) {
  const openedCards = yield select(selectOpenedNotGuessedCards);
  if (openedCards.length === 0) {
    yield put(triggerCard(action.payload.id));
  }
  if (openedCards.length === 1) {
    yield put(triggerCard(action.payload.id));
    yield race({
      task: call(validateCardsAsync),
      cancel: take(TRIGGER_CARD_ASYNC),
    });
  }
  if (openedCards.length > 1) {
    yield put({ type: VALIDATE_CARDS });
    yield put(triggerCard(action.payload.id));
  }
}

export default function* watchTriggerCardAsync() {
  yield takeEvery(TRIGGER_CARD_ASYNC, triggerCardAsync);
}
