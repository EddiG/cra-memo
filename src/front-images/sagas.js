import { takeLatest, select, call, put } from 'redux-saga/effects';
import { selectImagesSources } from './selectors';
import preload from '../utils/preloader';
import { PRELOAD_FRONT_IMAGES_REQUIRE } from './constants';
import * as actions from './actions';

export function* preloadImages() {
  const images = yield select(selectImagesSources);
  yield call(preload, images);
  yield put(actions.preloadImagesComplete());
}

export default function* watchPreloadImages() {
  yield takeLatest(PRELOAD_FRONT_IMAGES_REQUIRE, preloadImages);
}
