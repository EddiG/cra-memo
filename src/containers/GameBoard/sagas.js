import { delay } from 'redux-saga';
import { all, select, put, call, takeLatest } from 'redux-saga/effects';
import { updateCards, closeCards } from '../Cards/actions';
import { generateCards as generateNewCards } from './generators';
import { GENERATE_CARDS, REFRESH_CARDS } from './constants';

export function* generateCards() {
  const cards = yield select(generateNewCards);
  yield put(updateCards(cards));
}

export function* refreshCards() {
  yield put(closeCards());
  yield call(delay, 1000);
  yield call(generateCards);
}

export function* watchGenerateCards() {
  yield takeLatest(GENERATE_CARDS, generateCards);
}

export function* watchRefreshCards() {
  yield takeLatest(REFRESH_CARDS, refreshCards);
}

export default function* rootSaga() {
  yield all([watchGenerateCards(), watchRefreshCards()]);
}
