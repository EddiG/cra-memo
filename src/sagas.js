import { fork, put } from 'redux-saga/effects';
import gameBoardSagas from './containers/GameBoard/sagas';
import cardsSagas from './containers/Cards/sagas';
import preloadFrontImages from './front-images/sagas';
import { preloadImagesRequire } from './front-images/actions';
import { generateCards } from './containers/GameBoard/actions';

export default function* rootSaga() {
  yield fork(gameBoardSagas);
  yield fork(cardsSagas);
  yield fork(preloadFrontImages);
  yield put(generateCards());
  yield put(preloadImagesRequire());
}
