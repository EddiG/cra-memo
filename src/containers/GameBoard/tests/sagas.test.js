import { generateCards, refreshCards } from '../sagas';

const cards = [
  {
    id: 'card1',
    isOpened: false,
    isGuessed: false,
    frontImageId: 'frontImage1',
    backImageId: 'backImage1',
  },
  {
    id: 'card2',
    isOpened: false,
    isGuessed: false,
    frontImageId: 'frontImage2',
    backImageId: 'backImage1',
  },
];

describe('GameBoard saga', () => {
  describe('generateCards', () => {
    it('should dispatch the action UPDATE_CARDS with an array of cards', () => {
      const generator = generateCards();

      const selectCards = generator.next().value;
      expect(selectCards).toMatchSnapshot();

      const putCards = generator.next(cards).value;
      expect(putCards).toMatchSnapshot();
    });
  });
  describe('refreshCards', () => {
    const generator = refreshCards();
    it('should dispatch an action CLOSE_CARDS', () => {
      const putCloseCards = generator.next().value;
      expect(putCloseCards).toMatchSnapshot();
    });
    it('should wait until cards would be closed', () => {
      const callDelay = generator.next().value;
      expect(callDelay).toMatchSnapshot();
    });
    it('should call generateCards function', () => {
      const putGenerateCards = generator.next().value;
      expect(putGenerateCards).toMatchSnapshot();
    });
  });
});
