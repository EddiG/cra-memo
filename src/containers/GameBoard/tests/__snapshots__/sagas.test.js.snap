// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard saga generateCards should dispatch the action UPDATE_CARDS with an array of cards 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`GameBoard saga generateCards should dispatch the action UPDATE_CARDS with an array of cards 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "cards": Array [
          Object {
            "backImageId": "backImage1",
            "frontImageId": "frontImage1",
            "id": "card1",
            "isGuessed": false,
            "isOpened": false,
          },
          Object {
            "backImageId": "backImage1",
            "frontImageId": "frontImage2",
            "id": "card2",
            "isGuessed": false,
            "isOpened": false,
          },
        ],
      },
      "type": "UPDATE_CARDS",
    },
    "channel": null,
  },
}
`;

exports[`GameBoard saga refreshCards should call generateCards function 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`GameBoard saga refreshCards should dispatch an action CLOSE_CARDS 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {},
      "type": "CLOSE_CARDS",
    },
    "channel": null,
  },
}
`;

exports[`GameBoard saga refreshCards should wait until cards would be closed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
